compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,386 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/RemoveUnnecessaryTransposeNetPass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <vector>
+
+namespace
+{
+
+class TaggedShapeAnalyzer final
+{
+public:
+  /**
+   * @brief check 'Transpose-Reshape-Transpose' can be replaced by one 'Reshape'.
+   *
+   * NOTE:
+   *  - It tracks How each dimension of shape is transformed/moved by using 'tags'.
+   *    Tag records the initial order of each dimension.
+   *
+   *  - Tag is only attached for dimensions with non-1 value.
+   *
+   *  - For details, please refer analyze_(transpose, reshape) comments.
+   */
+  bool canRemoveTransposes(const luci::CircleTranspose *f_tr, const luci::CircleReshape *m_rs,
+                           const luci::CircleTranspose *b_tr);
+
+private:
+  void init_shape_with_tag(const luci::CircleNode *in_tensor);
+  bool analyze_transpose(const luci::CircleTranspose *transpose_node);
+  bool analyze_reshape(const luci::CircleReshape *reshape_node);
+  bool verify_tag() const;
+
+  struct Dim final
+  {
+    uint32_t value;
+    std::vector<uint8_t> tags;
+  };
+  using Shape = std::vector<Dim>;
+
+  const uint8_t _start_tag = 0;
+  Shape _shape;
+};
+
+/**
+ * @brief prepare _shape to simulate some operation(Transpose, Reshape)
+ *
+ * NOTE: Tag is assgigned by for non-1 dimensions.
+ *
+ * Example:
+ *
+ * - input:
+ *   in_tensor(rank=4, shape=(1, 7, 7, 448))
+ *
+ * - after:
+ *   _shape: (value=1, tags={}), (value=7, tags={0}), (value=7, tags={1}), (value=448, tags={2})
+ */
+void TaggedShapeAnalyzer::init_shape_with_tag(const luci::CircleNode *in_tensor)
+{
+  uint8_t tag = _start_tag;
+  _shape.clear();
+
+  for (uint32_t i = 0; i < in_tensor->rank(); i++)
+  {
+    TaggedShapeAnalyzer::Dim dim;
+    {
+      dim.value = in_tensor->dim(i).value();
+      if (dim.value != 1)
+        dim.tags.push_back(tag++);
+    }
+    _shape.push_back(dim);
+  }
+}
+
+/**
+ * @brief simulate 'Transpose' using _shape
+ *
+ * Example:
+ *
+ * - input:
+ *  transpose_node(perm=(0, 3, 1, 2))
+ *
+ * - before:
+ *   _shape: (value=1, tags={}), (value=7, tags={0}), (value=7, tags={1}), (value=448, tags={2})
+ *
+ * - after:
+ *   _shape: (value=1, tags={}), (value=448, tags={2}), (value=7, tags={0}), (value=7, tags={1})
+ * */
+bool TaggedShapeAnalyzer::analyze_transpose(const luci::CircleTranspose *transpose_node)
+{
+  const luci::CircleConst *perm_node = dynamic_cast<luci::CircleConst *>(transpose_node->perm());
```

Comment:
`perm_node` can be null. our analyzer tool will warn this.

Reply:
maybe you should check conditions in `remove_unnecessary_transpose` and use `must_cast`